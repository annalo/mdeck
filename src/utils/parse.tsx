/* eslint-disable */
import { Marpit } from "@marp-team/marpit";
import TaskList from "markdown-it-task-lists";

const RULES = [
  "blockquote_open",
  "code_block",
  "fence",
  "heading_open",
  "image",
  "link_open",
  "list_item_open",
  "paragraph_open",
  "table_open",
];

function injectLineNumber(md: any) {
  const { marpit_inline_svg_open } = md.renderer.rules;

  // Enable line sync by per slides
  md.renderer.rules.marpit_inline_svg_open = (tokens, idx, opts, env, self) => {
    const slide = tokens
      .slice(idx + 1)
      .find((t) => t.type === "marpit_slide_open");

    if (slide.map?.length) {
      tokens[idx].attrJoin("class", "src-line");
      tokens[idx].attrSet("data-line", slide.map[0]);
    }

    const renderer = marpit_inline_svg_open || self.renderToken;
    return renderer.call(self, tokens, idx, opts, env, self);
  };

  // Enables line sync per elements
  for (const rule of RULES) {
    const original = md.renderer.rules[rule];

    md.renderer.rules[rule] = (tokens, idx, options, env, self) => {
      const token = tokens[idx];

      if (token.map?.length) {
        token.attrJoin("class", "code-line");
        token.attrSet("data-line", token.map[0]);
      }

      const renderer = original || self.renderToken;
      return renderer.call(self, tokens, idx, options, env, self);
    };
  }
}

export function parse(markdown: string): string {
  const marpit = new Marpit({
    inlineSVG: true,
    markdown: { html: true, linkify: true, typographer: true },
  })
    .use(TaskList)
    .use(injectLineNumber);

  const { html }: { html: string } = marpit.render(markdown);
  return html;
}
