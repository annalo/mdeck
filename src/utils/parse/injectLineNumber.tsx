const RULES = [
  "blockquote_open",
  "code_block",
  "fence",
  "heading_open",
  "image",
  "link_open",
  "list_item_open",
  "paragraph_open",
  "table_open",
];

export function injectLineNumber(md: Record<string, any>) {
  const { marpitInlineSvgOpen } = md.renderer.rules;

  // Enable line sync by per slides
  // eslint-disable-next-line no-param-reassign
  md.renderer.rules.marpit_inline_svg_open = (tokens, idx, opts, env, self) => {
    const slide = tokens
      .slice(idx + 1)
      .find((t) => t.type === "marpit_slide_open");

    if (slide.map?.length) {
      tokens[idx].attrJoin("class", "src-line");
      tokens[idx].attrSet("data-line", slide.map[0]);
    }

    const renderer = marpitInlineSvgOpen || self.renderToken;
    return renderer.call(self, tokens, idx, opts, env, self);
  };

  // Enables line sync per elements
  RULES.forEach((rule) => {
    const original = md.renderer.rules[rule];
    // eslint-disable-next-line no-param-reassign
    md.renderer.rules[rule] = (tokens, idx, options, env, self) => {
      const token = tokens[idx];

      if (token.map?.length) {
        token.attrJoin("class", "code-line");
        token.attrSet("data-line", token.map[0]);
      }

      const renderer = original || self.renderToken;
      return renderer.call(self, tokens, idx, options, env, self);
    };
  });
}
