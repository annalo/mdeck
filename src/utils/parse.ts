import { Marpit } from "@marp-team/marpit";
import TaskList from "markdown-it-task-lists";
import { injectLineNumber } from "utils/parsePlugins/injectLineNumber";

export const parse = (markdown: string): HtmlArray => {
  const marpit = new Marpit({
    inlineSVG: true,
    markdown: { html: true, linkify: true, typographer: true },
  })
    .use(TaskList)
    .use(injectLineNumber);

  const { html }: { html: HtmlArray } = marpit.render(markdown, {
    htmlAsArray: true,
  });
  return html;
};
