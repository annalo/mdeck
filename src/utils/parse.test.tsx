import React from "react";
import { render } from "@testing-library/react";
import { parse } from "./parse";

describe("parse/1", () => {
  test("should parse markdown string into html string", () => {
    const markdown = "# Slide 1\n---\n## Slide 2\n* Bullet 1\n* Bullet 2";
    const htmlArray = parse(markdown);

    const { asFragment } = render(
      <div dangerouslySetInnerHTML={{ __html: htmlArray[0] }} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("should inject line number as 'data-line' attribute", () => {
    const markdown = "# Slide 1\n---\n## Slide 2";
    const htmlArray = parse(markdown);
    expect(htmlArray[0].match(/data-line/g)).toHaveLength(2);
    expect(htmlArray[1].match(/data-line/g)).toHaveLength(2);
  });

  test("should highlight syntax", () => {
    const markdown =
      "```js\nvar foo = function (bar) {\n  return bar++;\n};\n\nconsole.log(foo(5));\n```";
    const htmlArray = parse(markdown);

    const { asFragment } = render(
      <div dangerouslySetInnerHTML={{ __html: htmlArray[0] }} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
