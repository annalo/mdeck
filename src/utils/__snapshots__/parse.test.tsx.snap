// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse/1 should highlight syntax 1`] = `
<DocumentFragment>
  <div>
    <svg
      classname="code-line"
      data-line="0"
      data-marpit-svg=""
      viewBox="0 0 1280 720"
    >
      <foreignobject
        height="720"
        width="1280"
      >
        <section
          id="1"
        >
          

          <pre>
            <code
              class="language-js"
              classname="code-line"
              data-line="0"
            >
              <span
                class="hljs-keyword"
              >
                var
              </span>
               foo = 
              <span
                class="hljs-function"
              >
                <span
                  class="hljs-keyword"
                >
                  function
                </span>
                 (
                <span
                  class="hljs-params"
                >
                  bar
                </span>
                ) 
              </span>
              {
  
              <span
                class="hljs-keyword"
              >
                return
              </span>
               bar++;
};


              <span
                class="hljs-built_in"
              >
                console
              </span>
              .log(foo(
              <span
                class="hljs-number"
              >
                5
              </span>
              ));

            </code>
          </pre>
          

        </section>
        

      </foreignobject>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`parse/1 should parse markdown string into html string 1`] = `
<DocumentFragment>
  <div>
    <svg
      classname="code-line"
      data-line="0"
      data-marpit-svg=""
      viewBox="0 0 1280 720"
    >
      <foreignobject
        height="720"
        width="1280"
      >
        <section
          id="1"
        >
          

          <h1
            classname="code-line"
            data-line="0"
          >
            Slide 1
          </h1>
          

        </section>
        

      </foreignobject>
    </svg>
  </div>
</DocumentFragment>
`;
